### PR TITLE
Checks for connectivity when App is moving to foreground.

### PR DESCRIPTION
**Pull request by wu-hui**
Attempt to fix https://github.com/firebase/firebase-ios-sdk/issues/4905

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm [hunk 1]
```diff
@@ -107,9 +111,26 @@ explicit ConnectivityMonitorApple(
...
+
+#if TARGET_OS_IOS || TARGET_OS_TV
+    this->observer = [[NSNotificationCenter defaultCenter]
+        addObserverForName:UIApplicationWillEnterForegroundNotification
+                    object:nil
+                     queue:[NSOperationQueue mainQueue]
+                usingBlock:^(NSNotification* note) {
+                  SCNetworkReachabilityFlags flags;
+                  if (SCNetworkReachabilityGetFlags(reachability_, &flags)) {
+                    this->MaybeInvokeCallbacks(ToNetworkStatus(flags));
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm]: As discussed over chat, when the described scenario happens, we are getting a `kSCNetworkReachabilityFlagsConnectionRequired` which translates to `NetworkStatus::Unavailable`, then the grpc connection will be reset.

File: Firestore/core/src/firebase/firestore/remote/connectivity_monitor.h [hunk 2]
```diff
@@ -71,6 +71,9 @@ class ConnectivityMonitor {
   // Invokes callbacks only if the status changed.
   void MaybeInvokeCallbacks(NetworkStatus new_status);
 
+  // Invokes callbacks and set net status to `new_status`.
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/connectivity_monitor.h]: nit: change "set net status" to "sets the new status"

File: Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm [hunk 3]
```diff
@@ -107,9 +111,31 @@ explicit ConnectivityMonitorApple(
...
+    this->observer_ = [[NSNotificationCenter defaultCenter]
+        addObserverForName:UIApplicationWillEnterForegroundNotification
+                    object:nil
+                     queue:[NSOperationQueue mainQueue]
+                usingBlock:^(NSNotification* note) {
+                  SCNetworkReachabilityFlags flags{};
+                  if (SCNetworkReachabilityGetFlags(reachability_, &flags)) {
+                    auto status = ToNetworkStatus(flags);
+                    if (status != NetworkStatus::Unavailable) {
+                      this->InvokeCallbacks(status);
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm]: Add a comment here indicating what we're fixing. Something like:

> There may have been network changes while Firestore was in the background for which we did not get `OnReachabilityChangedCallback` notifications. If entering the foreground and we have a connection, reset the connection to ensure that RPCs don't have to wait for TCP timeouts.

Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm]: Internal callbacks need to run on the `AsyncQueue`. You might want to factor out an `OnEnteredForeground` member (and put it near `OnReachabilityChanged` to make this parallel clearer. Something like this:

```c++
void OnEnteredForeground() {
  SCNetworkReachabilityFlags flags{};
  if (!SCNetworkReachabilityGetFlags(reachability_, &flags)) return;

  queue()->Enqueue([this, flags] {
    auto status = ToNetworkStatus(flags);
    if (status != NetworkStatus::Unavailable) {
      this->InvokeCallbacks(status);
    } else {
      this->MaybeInvokeCallbacks(status);
    }
  });
}
```

File: Firestore/core/src/firebase/firestore/remote/connectivity_monitor.h (hunk 2 above)
Comment (wu-hui) [reply Firestore/core/src/firebase/firestore/remote/connectivity_monitor.h]: Done.

File: Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm (hunk 3 above)
Comment (wu-hui) [reply Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm]: Done.

Comment (wu-hui) [reply Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm]: Done.

**Pull request merged**